### PR TITLE
vet: add check for trailing spaces

### PR DESCRIPTION
**Pull request by purnesh42H**
RELEASE NOTES: None

### CODE REVIEW

**Comment by purnesh42H** [comment]
Will fix the existing files before getting this in

**Comment by dfawley** [comment]
> vet: add check for trailing spaces

Release notes are for users.  Only gRPC-Go developers will care about vet.  `None` is appropriate here.

File: scripts/vet.sh [hunk 1]
```diff
@@ -184,4 +184,11 @@ revive -formatter plain ./... >"${REV_OUT}" || true
 # TODO: Remove `|| true` to unskip linter failures once existing issues are fixed.
 (noret_grep -v "unused-parameter" "${REV_OUT}" | not grep -v "\.pb\.go:") || true
 
+# Collection of trailing spaces analysis checks
+TRAIL_SPACE_OUT="$(mktemp)"
```
Comment (dfawley) [inline scripts/vet.sh:188]: Let's use a pipe instead of creating a file whenever possible, please.

File: scripts/vet.sh [hunk 2]
```diff
@@ -184,4 +184,11 @@ revive -formatter plain ./... >"${REV_OUT}" || true
...
+find . -type f -exec grep -Hn "[[:blank:]]$" {} \; >"${TRAIL_SPACE_OUT}"
+
+# Error for anything other than in .git directory, vendor directory and *.md files.
+noret_grep -v "./.git" "${TRAIL_SPACE_OUT}" | noret_grep -v "vendor" | not grep -v ".md"
```
Comment (arjan-bal) [inline scripts/vet.sh:192]: Can we avoid scanning these files instead of excluding them later? `find` has a `-prune` flag that may be useful.

Comment (purnesh42H) [reply scripts/vet.sh]: Done

File: scripts/vet.sh (hunk 1 above)
Comment (purnesh42H) [reply scripts/vet.sh]: Done

**Comment by purnesh42H** [comment]
> Will fix the existing files before getting this in
> 
> ```
> ./xds/internal/test/e2e/run.sh:6:go test . 
> ./authz/grpc_authz_server_interceptors_test.go:60:			authzPolicy: `{		
> ./authz/grpc_authz_server_interceptors_test.go:62:				"allow_rules": 
> ./authz/grpc_authz_end2end_test.go:81:				"allow_rules": 
> ./authz/grpc_authz_end2end_test.go:169:							"headers": 
> ./authz/grpc_authz_end2end_test.go:173:									"values": 
> ./authz/grpc_authz_end2end_test.go:253:						"request": 
> ./security/advancedtls/testdata/server_cert_2.pem:52:            X509v3 Subject Key Identifier: 
> ./security/advancedtls/testdata/server_cert_2.pem:54:            X509v3 Authority Key Identifier: 
> ./security/advancedtls/testdata/server_cert_2.pem:57:            X509v3 Basic Constraints: 
> ./security/advancedtls/testdata/server_cert_2.pem:59:            X509v3 Key Usage: 
> ./security/advancedtls/testdata/client_cert_2.pem:52:            X509v3 Subject Key Identifier: 
> ./security/advancedtls/testdata/client_cert_2.pem:54:            X509v3 Authority Key Identifier: 
> ./security/advancedtls/testdata/client_cert_2.pem:57:            X509v3 Basic Constraints: 
> ./security/advancedtls/testdata/client_cert_2.pem:59:            X509v3 Key Usage: 
> ./security/advancedtls/testdata/server_cert_3.txt:36:            X509v3 Authority Key Identifier: 
> ./security/advancedtls/testdata/server_cert_3.txt:39:            X509v3 Basic Constraints: 
> ./security/advancedtls/testdata/server_cert_3.txt:41:            X509v3 Key Usage: 
> ./security/advancedtls/testdata/server_cert_3.txt:43:            X509v3 Subject Alternative Name: 
> ./security/advancedtls/testdata/server_cert_1.txt:52:            X509v3 Subject Key Identifier: 
> ./security/advancedtls/testdata/server_cert_1.txt:54:            X509v3 Authority Key Identifier: 
> ./security/advancedtls/testdata/server_cert_1.txt:57:            X509v3 Basic Constraints: 
> ./security/advancedtls/testdata/server_cert_1.txt:59:            X509v3 Key Usage: 
> ./security/advancedtls/testdata/server_cert_1.pem:52:            X509v3 Subject Key Identifier: 
> ./security/advancedtls/testdata/server_cert_1.pem:54:            X509v3 Authority Key Identifier: 
> ./security/advancedtls/testdata/server_cert_1.pem:57:            X509v3 Basic Constraints: 
> ./security/advancedtls/testdata/server_cert_1.pem:59:            X509v3 Key Usage: 
> ./security/advancedtls/testdata/server_cert_2.txt:52:            X509v3 Subject Key Identifier: 
> ./security/advancedtls/testdata/server_cert_2.txt:54:            X509v3 Authority Key Identifier: 
> ./security/advancedtls/testdata/server_cert_2.txt:57:            X509v3 Basic Constraints: 
> ./security/advancedtls/testdata/server_cert_2.txt:59:            X509v3 Key Usage: 
> ./security/advancedtls/testdata/server_cert_2.txt:91:         
> ./security/advancedtls/testdata/another_client_cert_1.pem:52:            X509v3 Subject Key Identifier: 
> ./security/advancedtls/testdata/another_client_cert_1.pem:54:            X509v3 Authority Key Identifier: 
> ./security/advancedtls/testdata/another_client_cert_1.pem:57:            X509v3 Basic Constraints: 
> ./security/advancedtls/testdata/another_client_cert_1.pem:59:            X509v3 Key Usage: 
> ./security/advancedtls/testdata/server_cert_localhost_1.pem:52:            X509v3 Subject Key Identifier: 
> ./security/advancedtls/testdata/server_cert_localhost_1.pem:54:            X509v3 Authority Key Identifier: 
> ./security/advancedtls/testdata/server_cert_localhost_1.pem:57:            X509v3 Basic Constraints: 
> ./security/advancedtls/testdata/server_cert_localhost_1.pem:59:            X509v3 Key Usage: 
> ./security/advancedtls/testdata/server_cert_localhost_1.pem:61:            X509v3 Subject Alternative Name: 
> ./security/advancedtls/testdata/client_cert_1.pem:52:            X509v3 Subject Key Identifier: 
> ./security/advancedtls/testdata/client_cert_1.pem:54:            X509v3 Authority Key Identifier: 
> ./security/advancedtls/testdata/client_cert_1.pem:57:            X509v3 Basic Constraints: 
> ./security/advancedtls/testdata/client_cert_1.pem:59:            X509v3 Key Usage: 
> ./examples/features/advancedtls/creds/server_cert_revoked.pem:52:            X509v3 Subject Key Identifier: 
> ./examples/features/advancedtls/creds/server_cert_revoked.pem:54:            X509v3 Authority Key Identifier: 
> ./examples/features/advancedtls/creds/server_cert_revoked.pem:56:            X509v3 Basic Constraints: 
> ./examples/features/advancedtls/creds/server_cert_revoked.pem:58:            X509v3 Key Usage: 
> ./examples/features/advancedtls/creds/server_cert_revoked.pem:60:            X509v3 Subject Alternative Name: 
> ./examples/features/advancedtls/creds/client_cert.pem:52:            X509v3 Subject Key Identifier: 
> ./examples/features/advancedtls/creds/client_cert.pem:54:            X509v3 Authority Key Identifier: 
> ./examples/features/advancedtls/creds/client_cert.pem:56:            X509v3 Basic Constraints: 
> ./examples/features/advancedtls/creds/client_cert.pem:58:            X509v3 Key Usage: 
> ./examples/features/advancedtls/creds/client_cert.pem:60:            X509v3 Subject Alternative Name: 
> ./examples/features/advancedtls/creds/server_cert.pem:52:            X509v3 Subject Key Identifier: 
> ./examples/features/advancedtls/creds/server_cert.pem:54:            X509v3 Authority Key Identifier: 
> ./examples/features/advancedtls/creds/server_cert.pem:56:            X509v3 Basic Constraints: 
> ./examples/features/advancedtls/creds/server_cert.pem:58:            X509v3 Key Usage: 
> ./examples/features/advancedtls/creds/server_cert.pem:60:            X509v3 Subject Alternative Name: 
> ./examples/features/advancedtls/creds/client_cert_revoked.pem:52:            X509v3 Subject Key Identifier: 
> ./examples/features/advancedtls/creds/client_cert_revoked.pem:54:            X509v3 Authority Key Identifier: 
> ./examples/features/advancedtls/creds/client_cert_revoked.pem:56:            X509v3 Basic Constraints: 
> ./examples/features/advancedtls/creds/client_cert_revoked.pem:58:            X509v3 Key Usage: 
> ./examples/features/advancedtls/creds/client_cert_revoked.pem:60:            X509v3 Subject Alternative Name: 
> ```

ALL FIXED in follow up commit

File: scripts/vet.sh [hunk 3]
```diff
@@ -184,4 +184,7 @@ revive -formatter plain ./... >"${REV_OUT}" || true
 # TODO: Remove `|| true` to unskip linter failures once existing issues are fixed.
 (noret_grep -v "unused-parameter" "${REV_OUT}" | not grep -v "\.pb\.go:") || true
 
+# Error if trailing spaces found in any files excluding files in .git directory and *.md files
+$(find . -path ./.git -prune -o -type f ! -name "*.md" -exec grep -Hn "[[:blank:]]$" {} \;) | fail_on_output
```
Comment (dfawley) [inline scripts/vet.sh:188]: Why exclude markdown files? I think trailing whitespaces should always be disallowed?

Comment (arjan-bal) [inline scripts/vet.sh:188]: Why surround the command in `$()`? This will cause the output of grep to be executed, which isn't what we want. I believe we can remove the surrounding `$()`
```
find . -path ./.git -prune -o -type f ! -name "*.md" -exec grep -Hn "[[:blank:]]$" {} \; | fail_on_output
```

However, this will not show all the violations because `fail_on_output` is implemented to exit early with a non-zero exit code if it sees any output. So it closes the read side of the pipe causing errors similar to the following if grep is still writing more results.
```sh
find . -path ./.git -prune -o -type f ! -name "*.md" -exec grep -Hn "[[:blank:]]$" {} \; | fail_on_output   
./features/advancedtls/creds/server_cert.pem:52:            X509v3 Subject Key Identifier: 
./features/advancedtls/creds/server_cert.pem:54:            X509v3 Authority Key Identifier: 
./features/advancedtls/creds/server_cert.pem:56:            X509v3 Basic Constraints: 
./features/advancedtls/creds/server_cert.pem:58:            X509v3 Key Usage: 
./features/advancedtls/creds/server_cert.pem:60:            X509v3 Subject Alternative Name: 
find: ‘grep’ terminated by signal 13
find: ‘grep’ terminated by signal 13
```
A workaround would be to define a version of `fail_on_output` that reads the entire input before exiting, the following code from an LLM demonstrates this:
```sh
fail_on_output() {                                                                                       
  output=$(tee /dev/stderr) # Read the entire output
  if [ -n "$output" ]; then
    return 1
  fi
}
```

File: scripts/vet.sh [hunk 4]
```diff
@@ -184,4 +184,7 @@ revive -formatter plain ./... >"${REV_OUT}" || true
 # TODO: Remove `|| true` to unskip linter failures once existing issues are fixed.
 (noret_grep -v "unused-parameter" "${REV_OUT}" | not grep -v "\.pb\.go:") || true
 
+# Error if trailing spaces found in any files excluding files in .git directory and *.md files
+find . -path ./.git -prune -o -type f -exec grep -Hn "[[:blank:]]$" {} \; | fail_on_output
```
Comment (dfawley) [inline scripts/vet.sh:188]: Does `git grep "[[:blank:]]$" | fail_on_output` not work?  That seems a lot simpler.

File: scripts/vet.sh (hunk 3 above)
Comment (purnesh42H) [reply scripts/vet.sh]: Yeah missed it. Good catch. The current command does show the violations though

Comment (purnesh42H) [reply scripts/vet.sh]: FIxed

File: scripts/vet.sh (hunk 4 above)
Comment (purnesh42H) [reply scripts/vet.sh:188]: yeah for some reason grep fails `fail_on_output` even when there are no violations of trailing whitespace. We probably need to modify `fail_on_output()` function or use something else

File: scripts/vet.sh [hunk 5]
```diff
@@ -184,4 +184,7 @@ revive -formatter plain ./... >"${REV_OUT}" || true
 # TODO: Remove `|| true` to unskip linter failures once existing issues are fixed.
 (noret_grep -v "unused-parameter" "${REV_OUT}" | not grep -v "\.pb\.go:") || true
 
+# Error if trailing spaces found in any files excluding files in .git directory and *.md files
```
Comment (arvindbr8) [inline scripts/vet.sh:187]:
```suggestion
# Error if trailing spaces found in any files excluding files in .git directory.
```

File: scripts/vet.sh (hunk 4 above)
Comment (arvindbr8) [reply scripts/vet.sh:188]: I gave it a try on my end, and it seems to be working fine

```console
~/grpc-go pr/purnesh42H/7576*
❯ git grep "[[:blank:]]$" | fail_on_output
~/grpc-go pr/purnesh42H/7576*
❯ echo $?
0

~/grpc-go pr/purnesh42H/7576*
❯ git grep "[[:blank:]]" | fail_on_output
~/grpc-go pr/purnesh42H/7576*
❯ echo $?
1
```

**Review by arvindbr8 — CHANGES_REQUESTED** [review]

File: scripts/vet.sh (hunk 4 above)
Comment (purnesh42H) [reply scripts/vet.sh:188]: but I don't i see Success at the end when running `scripts/vet.sh` with this command even though no violations. That's why I am not sure.

Comment (dfawley) [reply scripts/vet.sh:188]: See the other places we are doing `git grep` in here then...  E.g.

```sh
# - Ensure that the deprecated protobuf dependency is not used.
not git grep "\"github.com/golang/protobuf/*" -- "*.go" ':(exclude)reflection/test/grpc_testing_not_regenerate/*'
```

Comment (purnesh42H) [reply scripts/vet.sh]: "git grep '[[:blank:]]$' returns an exit status of 1 when no trailing spaces are found.. Hence, we have to negate it. Let me know if this looks good now

File: scripts/vet.sh [hunk 6]
```diff
@@ -183,4 +183,7 @@ revive -set_exit_status=1 -exclude "reflection/test/grpc_testing_not_regenerate/
 # TODO: Remove `|| true` to unskip linter failures once existing issues are fixed.
 (not grep -v "\.pb\.go:" "${REV_OUT}") || true
 
+# Ensure that no trailing spaces are found.
+not git grep '[[:blank:]]$'
```
Comment (arvindbr8) [inline scripts/vet.sh:187]: Optional: 

Should we group all the git grep together: near ~line:40

File: credentials/alts/internal/proto/grpc_gcp/handshaker.pb.go
Comment (arvindbr8) [inline credentials/alts/internal/proto/grpc_gcp/handshaker.pb.go:1]: we should probably do a separate PR for regenerate

File: internal/proto/grpc_lookup_v1/rls_config.pb.go
Comment (arvindbr8) [inline internal/proto/grpc_lookup_v1/rls_config.pb.go:1]: same, this one should be part of a separate PR

**Review by arvindbr8 — APPROVED** [review]
LGTM.  We should create a separate PR for all the proto regeneratation.

File: credentials/alts/internal/proto/grpc_gcp/handshaker.pb.go
Comment (purnesh42H) [reply credentials/alts/internal/proto/grpc_gcp/handshaker.pb.go:1]: vet proto check was failing if i didn't regenerate

File: internal/proto/grpc_lookup_v1/rls_config.pb.go
Comment (purnesh42H) [reply internal/proto/grpc_lookup_v1/rls_config.pb.go:1]: same: vet proto check was failing if i didn't regenerate

File: credentials/alts/internal/proto/grpc_gcp/handshaker.pb.go
Comment (arvindbr8) [reply credentials/alts/internal/proto/grpc_gcp/handshaker.pb.go:1]: The vet-proto is non-blocking, so I wouldnt worry to fix it as part of an unrelated PR.

File: internal/proto/grpc_lookup_v1/rls_config.pb.go
Comment (dfawley) [reply internal/proto/grpc_lookup_v1/rls_config.pb.go:1]: +1; please leave these changes out of this PR and send a separate one to fix vet-proto.  Since it's non-blocking, this can still be merged without these changes.

File: credentials/alts/internal/proto/grpc_gcp/handshaker.pb.go
Comment (purnesh42H) [reply credentials/alts/internal/proto/grpc_gcp/handshaker.pb.go:1]: ah ok. Yeah didn't know it was non blocking. Reverted regenerate

File: internal/proto/grpc_lookup_v1/rls_config.pb.go
Comment (purnesh42H) [reply internal/proto/grpc_lookup_v1/rls_config.pb.go:1]: Done

File: scripts/vet.sh (hunk 6 above)
Comment (purnesh42H) [reply scripts/vet.sh]: yeah moved to line 70 after other not git grep

File: scripts/vet.sh [hunk 7]
```diff
@@ -67,6 +67,9 @@ not git grep "\"github.com/golang/protobuf/*" -- "*.go" ':(exclude)reflection/te
 # - Ensure all usages of grpc_testing package are renamed when importing.
 not git grep "\(import \|^\s*\)\"google.golang.org/grpc/interop/grpc_testing" -- "*.go"
 
+# Ensure that no trailing spaces are found.
```
Comment (dfawley) [inline scripts/vet.sh:70]: I'm not sure why, exactly, but all the section comments start with `-`; please continue the pattern here.

**Review by dfawley — APPROVED** [review]
